webguitest/fabi/js/flipcomm.js

```diff
@@ -0,0 +1,527 @@
+        increaseProgress(10);
+        var percentPerSlot = 50 / thiz.getSlots().length;
+        var saveSlotsPromise = new Promise(function (resolve) {
+            loadAndSaveSlot(thiz.getSlots(), 0);
+
+            function loadAndSaveSlot(slots, i) {
+                if (i >= slots.length) {
+                    resolve();
+                    return;
+                }
+                var slot = slots[i];
+                thiz.testConnection().then(function () {
+                    loadSlotByConfig(slot);
+                    thiz.testConnection().then(function () {
+                        sendAtCmdNoResultHandling('AT SA ' + slot);
+                        increaseProgress(percentPerSlot);
+                        loadAndSaveSlot(slots, i + 1);
+                    });
+                });
+            }
+        });
+
+        function increaseProgress(percent) {
+            progress += percent;
+            updateProgressHandler(progress);
+        }
+
+        return new Promise(function (resolve) {
+            saveSlotsPromise.then(function () {
+                thiz.testConnection().then(function () {
+                    increaseProgress(30);
+                    loadSlotByConfig(_currentSlot).then(function () {
+                        increaseProgress(10);
+                        thiz.resumeLiveValueListener();
+                        resolve();
+                    });
+                });
+            });
+        });
+    };
+
+    thiz.calibrate = function () {
+        sendAtCmdNoResultHandling('AT CA');
+        return thiz.testConnection();
+    };
+
+    thiz.rotate = function () {
+        var currentOrientation = thiz.getConfig(thiz.ORIENTATION_ANGLE);
+        thiz.setValue(thiz.ORIENTATION_ANGLE, (currentOrientation + 90) % 360, 0);
+        sendAtCmdNoResultHandling('AT CA');
+        return thiz.testConnection();
+    };
+
+    thiz.startLiveValueListener = function (handler) {
+        console.log('starting listening to live values...');
+        setLiveValueHandler(handler);
+    };
+
+    thiz.stopLiveValueListener = function () {
+        setLiveValueHandler(null);
+        console.log('listening to live values stopped.');
+    };
+
+    thiz.pauseLiveValueListener = function () {
+        thiz.sendATCmd('AT ER');
+        console.log('listening to live values stopped.');
+    };
+
+    thiz.resumeLiveValueListener = function () {
+        if (_valueHandler) {
+            thiz.sendATCmd('AT SR');
+            console.log('listening to live values resumed.');
+        } else {
+            console.log('listening to live values not resumed, because no value handler.');
+        }
+    };
+
+    // return 'true' if SipAndPuff device is available, else 'false'
+    thiz.isSipAndPuffLive = function () {
+        return new Promise(
+            function(resolve, reject) {
+		thiz.startLiveValueListener(function (livedata) {
+		    if (livedata) {
+			resolve(true);
+		    }
+		    else {
```

Comment:
I don't think you have to check for `if (livedata)` here because in any case the listener function is called, live data will be present. Instead of that you could use a `setTimeout()` and if the listener function is not called after a specific amount of time (e.g. 2 seconds), you could call `thiz.stopLiveValueListener()` and `resolve(false)`.

Reply:
okay

Reply:
`thiz.startLiveValueListener(fn);` starts listening to the live values from the flipmouse and the passed function `fn` is called, if live values arrive from the flipmouse. MockCommunicator is used to provide fake "live" values in the demo mode, so this piece of code we are talking about should not behave differently, if the "live" values are fake and not coming from the real flipmouse.

what I meant was something like:
```
    thiz.isSipAndPuffLive = function () {
        return new Promise(
            function(resolve, reject) {
		thiz.startLiveValueListener(function () {
		     thiz.stopLiveValueListener();
                     resolve(true);
                 }
                 setTimeout(function() { thiz.stopLiveValueListener(); resolve(false); }, 2000);
            }
       )};
```
if the callback function of `thiz.startLiveValueListener` (passed as parameter) is ever called, we know that we are receiving live values and can `resolve(true)`. Parallel we start a timer that `resolves(false)` after 2 seconds, if no data was received.

Reply:
Thanks for the details. But in this demo mode 'Sip and Puff' menu button is always displayed because of mock data. Will you be removing the function sending mock data in actual mode?

Reply:
I think it's ok that the Sip and Puff menu is shown in the demo/mock mode.

At https://github.com/asterics/FLipMouse-esp32/blob/master/webguitest/spiffs_content/js/flipcomm.js#L385 you can see that it is tried to open a websocket to the actual FlipMouse and if this fails then the mock mode is activated. So in the real mode no mock data will be sent.